dl/coordinator: admin endpoint for getting coordinator state

andrwng opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->
We have some upcoming breaking Iceberg Topic schema changes that will require some manual intervention to upgrade to.

Existing Iceberg Topics will need to:

1. pause translation by disabling Iceberg for the affected topics
2. upgrade
3. quiesce pending coordinator state (wait for commits to happen, or reset explicitly with future tooling)
4. perform some manual data rewriting of Iceberg data to be compatible with the new schema
5. reenable Iceberg with the upgraded schema mapping

This endpoint in particular will be used in step 2.

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

* None
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

andrwng committed:
dl/coordinator: add get_topic_state

Adds an RPC endpoint to the coordinator leader that fetches topic
states. This will be used by an upcoming endpoint to be able to get
visibility into the coordinator state (e.g. to allow users to know if
there are any pending files remaining to be committed).

andrwng committed:
admin: datalake protobuf spec for get_state

Adds protobuf definitions for getting the datalake coordinator state as
an admin endpoint. This will be an important endpoint for upcoming
upgrades, where some breaking changes will occur to Iceberg tables.

Existing Iceberg Topics will need to:
1. pause translation
2. quiesce pending state before upgrading
3. perform some manual data rewriting
4. upgrade
5. reenable Iceberg with the upgraded schema mapping.

This endpoint in particular will be used in step 2.

I considered adding a more specialized endpoint to check if there is any
pending state (i.e. return a simple boolean), but thought the visibility
of seeing all the topic state would be useful in practice.

andrwng committed:
admin: datalake service for getting topic state

Adds the newly defined admin protobuf RPC endpoints as a service. The
implementation collects the input topics and sends them to the
appropriate coordinator partition.

andrwng committed:
rptest: expose datalake endpoint in admin v2

andrwng committed:
rptest: test for datalake endpoint

Adds a ducktape test for the newly added admin endpoint to get the topic
state from the datalake coordinator.

Added this test without type checking (similar to the other datalake
tests) because it relies on datalake primitives that are not
type-check-friendly.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/74360-->
<!--0ee201fc-e3ed-4f31-ad5b-5f561e97ec33-->
<!--title-->
## Retry command for [Build#74360](https://buildkite.com/redpanda/redpanda/builds/74360)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/tiered_storage_model_test.py::TieredStorageTest.test_tiered_storage@{"cloud_storage_type_and_url_style":[1,"virtual_host"],"test_case":{"name":"(TS_Read == True, TS_Timequery == True, SpilloverManifestUploaded == True)"}}
```
<!--/body-->

oleiman commented on src/v/redpanda/admin/services/datalake/datalake.cc:
```diff
@@ -0,0 +1,177 @@
+    // Group topics by coordinator partition.
+    chunked_hash_map<model::partition_id, chunked_vector<model::topic>>
+      topics_by_partition;
+    for (const auto& topic_name : req.get_topics()) {
```
it occurs to me that it's probably not uncommon to want "all topics tracked by the coordinator". should support that at the API level?

andrwng replied:
Yeah makes sense, done

oleiman replied:
in the commit at the end? (this is a note for me, sorry in advance if I forget to delete it)

oleiman reviewed (commented):
seems legit. one question about common usage

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--70873eaf-d84e-4a89-8f07-58a9aee96371-->
<!--TEST RESULTS-->
<!--7a4983d2-9130-4cf2-be6d-211aa27f1140-->
<!--TEST RESULTS-->
<!--39d09461-358c-4e94-95a5-688cd0241add-->
<!--TEST RESULTS-->
<!--f783fafc-2c70-4d1b-b53b-543710520cb8-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74360">build#74360</a></summary>


| test_class                          | test_method                                  | test_arguments                                                                                                                                                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                             |
|:------------------------------------|:---------------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DataMigrationsApiTest               | test_creating_and_listing_migrations         | null                                                                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/74360#0199efc9-e52d-4779-9c6a-3f650fb0367f | FLAKY         | 19/21    | upstream reliability is '98.90510948905109'. current run reliability is '90.47619047619048'. drift is 8.42892 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DataMigrationsApiTest&test_method=test_creating_and_listing_migrations |
| SegmentMsTest                       | test_segment_rolling_with_retention_consumer | null                                                                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/74360#0199efc5-c4d5-48c6-b060-4c29c94450d4 | FLAKY         | 19/21    | upstream reliability is '90.28459273797841'. current run reliability is '90.47619047619048'. drift is -0.1916 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=SegmentMsTest&test_method=test_segment_rolling_with_retention_consumer |
| NodesDecommissioningTest            | test_recommissioning_node                    | {"cloud_topic": false}                                                                                                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/74360#0199efc5-c4d1-4276-b30d-a4c8dd08db15 | FLAKY         | 19/21    | upstream reliability is '97.11538461538461'. current run reliability is '90.47619047619048'. drift is 6.63919 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_recommissioning_node         |
| RedpandaNodeOperationsSmokeTest     | test_node_ops_smoke_test                     | {"cloud_storage_type": 1, "mixed_versions": false}                                                                                                             | integration | https://buildkite.com/redpanda/redpanda/builds/74360#0199efc9-e52c-4ea5-b812-70b9c874fed9 | FLAKY         | 18/21    | upstream reliability is '100.0'. current run reliability is '85.71428571428571'. drift is 14.28571 and the allowed drift is set to 50. The test should PASS            | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RedpandaNodeOperationsSmokeTest&test_method=test_node_ops_smoke_test   |
| RedpandaNodeOperationsSmokeTest     | test_node_ops_smoke_test                     | {"cloud_storage_type": 1, "mixed_versions": true}                                                                                                              | integration | https://buildkite.com/redpanda/redpanda/builds/74360#0199efc9-e52d-4779-9c6a-3f650fb0367f | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RedpandaNodeOperationsSmokeTest&test_method=test_node_ops_smoke_test   |
| TieredStorageTest                   | test_tiered_storage                          | {"cloud_storage_type_and_url_style": [1, "virtual_host"], "test_case": {"name": "(TS_Read == True, TS_Timequery == True, SpilloverManifestUploaded == True)"}} | integration | https://buildkite.com/redpanda/redpanda/builds/74360#0199efc5-c4d1-4276-b30d-a4c8dd08db15 | FAIL          | 0/1      | The test has failed across all retries                                                                                                                                 | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TieredStorageTest&test_method=test_tiered_storage                      |
| WriteCachingFailureInjectionE2ETest | test_crash_all                               | {"use_transactions": false}                                                                                                                                    | integration | https://buildkite.com/redpanda/redpanda/builds/74360#0199efc5-c4d7-4124-a0d7-e404b00e304e | FLAKY         | 17/21    | upstream reliability is '81.7910447761194'. current run reliability is '80.95238095238095'. drift is 0.83866 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all         |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74382">build#74382</a></summary>


| test_class                        | test_method                        | test_arguments                                                                                                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                                |
|:----------------------------------|:-----------------------------------|:---------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkBasicTests              | test_topic_properties_replicated   | {"source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}} | integration | https://buildkite.com/redpanda/redpanda/builds/74382#0199f15b-cea1-4706-94e7-d32ee2056498 | FLAKY         | 18/21    | upstream reliability is '95.50561797752809'. current run reliability is '85.71428571428571'. drift is 9.79133 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkBasicTests&test_method=test_topic_properties_replicated         |
| DatalakeAdminEndpointTest         | test_coordinator_get_state_pending | {"catalog_type": "rest_jdbc", "cloud_storage_type": 1}                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/74382#0199f15b-1349-4bde-91f4-a914f476b454 | FLAKY         | 9/21     | The test was found to be new, and no failures are allowed                                                                                                              | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeAdminEndpointTest&test_method=test_coordinator_get_state_pending  |
| DatalakeAdminEndpointTest         | test_coordinator_get_state_pending | {"catalog_type": "rest_jdbc", "cloud_storage_type": 1}                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/74382#0199f15b-ce9f-4169-928d-0cf64ae19111 | FLAKY         | 13/21    | The test was found to be new, and no failures are allowed                                                                                                              | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeAdminEndpointTest&test_method=test_coordinator_get_state_pending  |
| DatalakeClusterRestoreTest        | test_slow_tiered_storage_dlq       | {"catalog_type": "nessie", "cloud_storage_type": 1}                                                            | integration | https://buildkite.com/redpanda/redpanda/builds/74382#0199f15b-1345-46f2-89de-886e85401a36 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeClusterRestoreTest&test_method=test_slow_tiered_storage_dlq       |
| VerifyConsumerOffsetsThruUpgrades | test_consumer_group_offsets        | {"versions_to_upgrade": 2}                                                                                     | integration | https://buildkite.com/redpanda/redpanda/builds/74382#0199f15b-ce9b-4157-9d1d-b07a1b3be6cd | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=VerifyConsumerOffsetsThruUpgrades&test_method=test_consumer_group_offsets |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74448">build#74448</a></summary>


| test_class           | test_method                     | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                       |
|:---------------------|:--------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkBasicTests | test_rapid_shadow_link_toggling | null             | integration | https://buildkite.com/redpanda/redpanda/builds/74448#0199f41f-c66d-4d4c-bb5e-d994abd6596b | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkBasicTests&test_method=test_rapid_shadow_link_toggling |
| ShadowLinkBasicTests | test_rapid_shadow_link_toggling | null             | integration | https://buildkite.com/redpanda/redpanda/builds/74448#0199f41f-e4d0-47b7-b98a-faa43ca4db0c | FLAKY         | 18/21    | upstream reliability is '92.65175718849841'. current run reliability is '85.71428571428571'. drift is 6.93747 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkBasicTests&test_method=test_rapid_shadow_link_toggling |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74469">build#74469</a></summary>


| test_class                    | test_method                    | test_arguments                                                                                                                                                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                               |
|:------------------------------|:-------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkingReplicationTests | test_replication_with_failures | null                                                                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/74469#0199f5aa-f8e9-492f-be58-c94f4e1aaa0c | FLAKY         | 10/21    | upstream reliability is '75.27405602923264'. current run reliability is '47.61904761904761'. drift is 27.65501 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_with_failures |
| ShadowLinkingReplicationTests | test_replication_with_failures | null                                                                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/74469#0199f5ab-92cb-41b0-9b2f-60ea48227a7e | FLAKY         | 6/21     | upstream reliability is '75.27405602923264'. current run reliability is '28.57142857142857'. drift is 46.70263 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_with_failures |
| TieredStorageTest             | test_tiered_storage            | {"cloud_storage_type_and_url_style": [1, "virtual_host"], "test_case": {"name": "(TS_Read == True, TS_Timequery == True, SpilloverManifestUploaded == True)"}} | integration | https://buildkite.com/redpanda/redpanda/builds/74469#0199f5aa-f8e7-45e2-906f-5d8b7e4f7c34 | FAIL          | 0/1      |                                                                                                                                                                         | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TieredStorageTest&test_method=test_tiered_storage                        |
| TimeQueryTest                 | test_timequery                 | {"batch_cache": false, "cloud_storage": true, "spillover": true}                                                                                               | integration | https://buildkite.com/redpanda/redpanda/builds/74469#0199f5ab-92c8-4ee7-8f64-f4fef69787f9 | FLAKY         | 20/21    | upstream reliability is '99.76415094339622'. current run reliability is '95.23809523809523'. drift is 4.52606 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TimeQueryTest&test_method=test_timequery                                 |
</details>
<!--/body-->

andrwng committed:
dl/coordinator: make CoordinatorGetState return all by default

Changes the `topics` argument to a `topics_filter`, where if it's empty
it returns all topics, and otherwise returns just what is in the filter.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/74382-->
<!--7abfa333-4cb8-4ddb-81dc-2be0edd74da7-->
<!--88521f6d-a38a-486c-86f9-a5538b3c0b56-->
<!--title-->
## Retry command for [Build#74382](https://buildkite.com/redpanda/redpanda/builds/74382)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/datalake/admin_endpoint_test.py::DatalakeAdminEndpointTest.test_coordinator_get_state_pending@{"catalog_type":"rest_jdbc","cloud_storage_type":1}
```
<!--/body-->

nvartolomei commented:
v cool! 👍

rockwotj commented on proto/redpanda/core/admin/v2/datalake.proto:
```diff
@@ -0,0 +1,83 @@
+
+syntax = "proto3";
+
+package redpanda.core.admin.v2;
```
Are we sure we want this public API? The API requirements are higher, it's fine but just want to ask the question

andrwng replied:
I'm on the fence about this -- on the one hand having it be internal would probably make our lives easier for evolving this, but this is also something that we'll need to call out for upgrades, so it felt right to make it public. I could go either way.

andrwng replied:
Chatted a bit more with storage folks; I think I'll move this into the debug endpoint just to be safe.

I don't think coordinator state will be evolved drastically over time so it should be pretty stable. That said, we definitely don't _need_ to expose all its guts like this just for this upgrade process (like maybe a subset could be public, longer term), and it feels a bit better to not lock ourselves in making it public.

rockwotj replied:
I agree it's safer to start internal because we can't yank back public. I would say just move this proto into the internal directory and that is all you should need to do

rockwotj commented on proto/redpanda/core/admin/v2/datalake.proto:
```diff
@@ -0,0 +1,83 @@
+
+option (redpanda.core.pbgen.cpp_namespace) = "proto::admin";
+
+service DatalakeService {
```
Please add a comment explaining what Datalake is, for the docs team, cloud team and customers.

See documentation guidelines here: https://redpandadata.slack.com/archives/C09C27KTCPK/p1759766098707339

andrwng replied:
Done

rockwotj commented on proto/redpanda/core/admin/v2/datalake.proto:
```diff
@@ -0,0 +1,83 @@
+
+service DatalakeService {
+    // Gets the coordinator state for the given topics.
+    rpc CoordinatorGetState(CoordinatorGetStateRequest)
```
GetCoordinatorState (and rename the request/response messages too)

andrwng replied:
Done

rockwotj commented on proto/redpanda/core/admin/v2/datalake.proto:
```diff
@@ -0,0 +1,83 @@
+    repeated string topics_filter = 1;
+}
+
+message CoordinatorGetStateResponse {
```
A GetCoordinatorStateResponse should return a single field of a message type CoordinatorState

andrwng replied:
Done

rockwotj commented on proto/redpanda/core/admin/v2/datalake.proto:
```diff
@@ -0,0 +1,83 @@
+    int64 last_offset = 2;
+    repeated DataFile files = 3;
+    repeated DataFile dlq_files = 4;
+    uint64 kafka_bytes_processed = 5;
```
The unit should be last: kafka_processed_bytes

andrwng replied:
Done

rockwotj commented on proto/redpanda/core/admin/v2/datalake.proto:
```diff
@@ -0,0 +1,83 @@
+    int64 revision = 1;
+    map<int32, PartitionState> partition_states = 2;
+    LifecycleState lifecycle_state = 3;
+    uint64 total_kafka_bytes_processed = 4;
```
unit goes last: total_kafka_processed_bytes

andrwng replied:
Done

rockwotj reviewed (commented):
Just reviewed the proto: we need to make sure we are adding comments to explain what these services, messages and RPCs are and what they are for. This is a solid start, nice work!

rockwotj commented on proto/redpanda/core/admin/v2/datalake.proto:
```diff
@@ -0,0 +1,83 @@
+message TranslatedOffsetRange {
+    int64 start_offset = 1;
+    int64 last_offset = 2;
+    repeated DataFile files = 3;
```
maybe data_files to more clearly separate it from dlq_files?

andrwng replied:
Done

andrwng committed:
admin: make datalake endpoint internal

...and some other feedback.

The history here got squashed a bit while making the datalake endpoint
internal, so there are some other changes lumped into this commit (it's
kind of a pain to untangle, especially with the generated code).

Other minor changes:
- added some comments to the protobuf
- CoordinatorGetState -> GetCoordinatorState
- made GetCoordinatorState return a CoordinatorState message that
  contains the topics state
- *kafka_bytes_processed -> *kafka_processed_bytes
- files -> data_files
- tweak to admin_endpoint_test.py to retry on admin endpoint error (only
  for the test case that attempts to hit the coordinator immediately)

andrwng commented:
Pushed a [new commit](https://github.com/redpanda-data/redpanda/pull/28084/commits/b543da0b020c7393e332bd09a436a12c8bd5ad27) at the end to address feedback (sorry, once I moved the proto to /internal, it became tricky to keep the history straight, but there aren't any major changes anyway)

andrwng commented:
Latest force push is to add back the datalake.proto file, accidentally removed when making the endpoint internal

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/74448-->
<!--f5ff5056-e2aa-4b53-8c9f-0d0fc4c68dbc-->
<!--4758d12c-5ab1-4f2b-b159-c764fa772565-->
<!--title-->
## Retry command for [Build#74448](https://buildkite.com/redpanda/redpanda/builds/74448)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/cluster_linking_e2e_test.py::ShadowLinkBasicTests.test_rapid_shadow_link_toggling
```
<!--/body-->

oleiman commented on src/v/redpanda/admin/services/datalake/datalake.cc:
```diff
@@ -0,0 +1,177 @@
+    for (const auto& [partition_id, topics] : topics_by_partition) {
+        datalake::coordinator::get_topic_state_request fe_req{
+          partition_id, topics.copy()};
+        auto fe_res = co_await _coordinator_fe->local().get_topic_state(
+          std::move(fe_req));
```
q: there are retries built in to the coordinator frontend, right?

andrwng replied:
Nah, only routing to the leader is done by the frontend. I'd considered adding retries in this PR, but opted to keep this PR slim

oleiman commented on proto/redpanda/core/admin/v2/datalake.proto:
```diff
@@ -32,7 +32,9 @@ service DatalakeService {
-    repeated string topics = 1;
+    // Names of topics whose state to return. If empty, returns state for all
+    // topics.
+    repeated string topics_filter = 1;
```
👍

oleiman commented on tests/rptest/tests/datalake/admin_endpoint_test.py:
```diff
@@ -76,14 +76,25 @@ def test_coordinator_get_state(
                 admin.datalake().coordinator_get_state(request)
             )
+            assert len(response.topic_states) == 1, (
+                f"Response should contain 2 topics: {response.topic_states}"
```
nitpick:

```suggestion
                f"Response should contain 1 topic: {response.topic_states}"
```

andrwng replied:
Will update if I end up pushing again

oleiman reviewed (approved):
lgtm
